activation/activation.go

```diff
@@ -240,53 +239,92 @@ func (b *Builder) StopSmeshing(deleteFiles bool) error {
 	b.smeshingMutex.Lock()
 	defer b.smeshingMutex.Unlock()
 
-	if !b.started.Load() {
+	if !b.started {
 		return errors.New("not started")
 	}
 
 	b.stop()
 	err := b.eg.Wait()
+	b.started = false
 	switch {
 	case err == nil || errors.Is(err, context.Canceled):
 		if !deleteFiles {
 			return nil
 		}
-
-		if err := discardBuilderState(b.nipostBuilder.DataDir()); err != nil && !errors.Is(err, fs.ErrNotExist) {
-			b.log.With().Error("failed to delete builder state", log.Err(err))
-			return err
-		}
-		if err := discardNipostChallenge(b.nipostBuilder.DataDir()); err != nil && !errors.Is(err, fs.ErrNotExist) {
-			b.log.With().Error("failed to delete nipost challenge", log.Err(err))
+		if err := nipost.RemoveChallenge(b.localDB, b.signer.NodeID()); err != nil {
+			b.log.With().Error("failed to remove nipost challenge", log.Err(err))
 			return err
 		}
-		if err := discardPost(b.nipostBuilder.DataDir()); err != nil && !errors.Is(err, fs.ErrNotExist) {
-			b.log.With().Error("failed to delete post", log.Err(err))
+		if err := discardBuilderState(b.nipostBuilder.DataDir()); err != nil && !errors.Is(err, fs.ErrNotExist) {
+			b.log.With().Error("failed to delete builder state", log.Err(err))
 			return err
 		}
-
 		return nil
 	default:
 		return fmt.Errorf("failed to stop smeshing: %w", err)
 	}
 }
 
+func (b *Builder) MovePostToDb() error {
+	post, err := loadPost(b.nipostBuilder.DataDir())
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		return nil // no post file, nothing to do
+	case err != nil:
+		return fmt.Errorf("loading post: %w", err)
+	default:
+	}
+
+	meta, err := initialization.LoadMetadata(b.nipostBuilder.DataDir())
+	if err != nil {
+		return fmt.Errorf("getting post info: %w", err)
+	}
+	commitmentAtxId := types.BytesToATXID(meta.CommitmentAtxId)
+
+	ch := &types.NIPostChallenge{
+		InitialPost:   post,
+		CommitmentATX: &commitmentAtxId,
+	}
+	if err := nipost.AddChallenge(b.localDB, b.signer.NodeID(), ch); err != nil {
+		return fmt.Errorf("adding post to db: %w", err)
+	}
```

Comment:
I'm not sure about keeping the initial POST as a part of `NipostChallenge`. I think it's unreliable because it is easy to accidentally wipe out the initial POST proof. I think this can already happen on a few paths when the nipostBuilder returns the `ErrATXChallengeExpired` error.

Reply:
In my opinion the problem here is that we return `ErrATXChallengeExpired` when we want to clear the challenge instead of just doing that with `nipost.RemoveChallenge` immediately. It feels like a hidden goto.

I think we shouldn't handle the initial PoST any different from all the other posts we do. We could even go as far as never deleting a challenge but just updating it when it becomes stale. In any case where `ErrATXChallengeExpired` is returned the only things that need to be updated in the challenge are the `PositioningATX` and the `PublishEpoch`, all other fields (must) stay the same.

Reply:
I separated the initial post from the challenge in the local DB. The initial post is only cleared on successful publication, the challenge on successful publication and when it expires.